Data-Structures/Utility-Collections.md

```diff
@@ -0,0 +1,260 @@
+
+Vector (and Array) variants are generally designed to allow (near) constant-time addressing of individual elements, whereas a standard List type may be most efficient when dealing with one end.
+
+Implementations:
+
+* Java: Vector, ArrayList, LinkedList, CopyOnWriteArrayList, LinkedHashSet
+* Scala: List, Vector
+* Clojure: list, vector
+* JavaScript: Array, TypedArray, Set (no duplicates)
+* C++: array (fixed-size), vector, list, forward_list
+
+##### `Queue`
+
+A FIFO list allowing only insertions in one end and retrieval from the other.
+
+Operations include:
+
+* `Add(element)`
+* `Remove()`
+* `Peek()`
+
+Implementations:
+
+* Java: Queue, Deque (double-ended)
+* Scala: Queue
+* C++: queue, deque (double-ended)
+
+##### `Stack`
+
+A LIFO list allowing only insertions and retrieval from a single end.
+
+Operations include:
+
+* `Push()`
+* `Pop()`
+* `Peek()`
+
+Implementations:
+
+* Java: Stack
+* Scala: Stack
+* C++: stack
+
+#### Ordered by comparator
+
+##### `SortedSet`
+
+A simple collection of non-duplicate entries that is ordered according to some comparison operation. 
+
+Operations include all those for an unordered Set and may extend to include:
+
+* `First()`
+* `Last()`
+* `Floor()?`
+* `Ceiling()?`
+* `Iterate()`
+* `SubSet()?`
+
+The `Iterate()` and `SubSet()?` operations may include "floor", "ceiling" and direction modifiers for range queries.
+
+Implementations:
+
+* Java: TreeSet, PriorityQueue
+* Scala: TreeSet
+* Clojure: sorted-set
+* C++: set, multiset (can contain duplicate elements)
+
+### Associative collections
+
+#### Unordered
+
+##### `Map`
+
+An unordered collection of key/value entries where, for any key, it contains zero or one value.
+
+Operations include:
+
+* `Set(key, value)`
+* `Get(key)`
+* `Contains(key)?`
+* `EntriesIterator()?`
+* `KeysIterator()?`
+* `ValuesIterator()?`
+* `Size()?`
+
+Implementations:
+
+* Java: HashMap
+* Scala: Map, HashMap
+* Clojure: hash-map
+* C++: unordered_map
+
+#### Ordered by construction
+
+##### `ListMap`
+
+An instance of a Map whose ordering is dictated by the insertion order of elements.
+
+Allows for all operations of a Map but the `*Iterator()` operations yield elements in insertion order.
+
+Implementations:
+
+* Java: LinkedHashMap
+* Scala: ListMap
+* Clojure: array-map
+* JavaScript: Map
+
+#### Ordered by comparator
+
+##### `SortedMap`
+
+A Map whose entries are ordered according to some comparison operation on the key.
+
+Allows for all operations of a Map but the `*Iterator()?` operations yield entries ordered by key (according to some comparison operation). In addition, a SortedMap can support forms of:
+
+* `First()?`
+* `Last()?`
+* `Floor()?`
+* `Ceiling()?`
+
+A `SubMap()?` is possible for generating an instance or view of a subset of the elements. `*Iterator()?` operations and `SubMap()?` may include floor, ceiling and direction modifiers for range queries.
```

Comment:
Can we just add new iterator operations for ranges and floor/ceiling w/ these new modifiers.

Depending on the implementation of the data structure I could see wanting either 1. a KeyValueIterator sorted by the sorting algorithm of the data structure **or** 2. a KeyValueIterator sorted by the fastest traversal possible, effectively being unordered.

I also really appreciate that we’re defining each type based on its **capabilities**. If we limit/remove any implied inheritance it makes it a lot easier to create collections that support capabilities of several collections that aren’t strictly heirachical in the way they inherit.